relay-quotas/src/global.rs

```diff
@@ -0,0 +1,256 @@
+#[derive(Clone, Copy, Hash)]
+struct BudgetKeyRef<'a> {
+    prefix: &'a str,
+    window: u64,
+}
+
+impl<'a> BudgetKeyRef<'a> {
+    fn new(quota: &'a RedisQuota) -> Self {
+        Self {
+            prefix: quota.prefix(),
+            window: quota.window(),
+        }
+    }
+}
+
+impl hashbrown::Equivalent<BudgetKey> for BudgetKeyRef<'_> {
+    fn equivalent(&self, key: &BudgetKey) -> bool {
+        self.prefix == key.prefix && self.window == key.window
+    }
+}
+
+struct SlottedCounter {
+    slot: usize,
+    counter: Counter,
+}
+
+impl SlottedCounter {
+    pub fn new() -> Self {
+        Self {
+            slot: 0,
+            counter: Counter::new(),
+        }
+    }
+
+    pub fn is_rate_limited(
+        &mut self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+        limit: u64,
+    ) -> Result<bool, RedisError> {
+        let current_slot = current_slot(quota.window());
+
+        match self.slot.cmp(&current_slot) {
+            Ordering::Greater => {
+                // TODO double check logic here
+                // in theory this should never happen only if someone messes with the system time
+                // be safe and dont rate limit
+                relay_log::error!("budget slot ahead of current slot");
+                Ok(false)
+            }
+            Ordering::Less => {
+                self.counter = Counter::new();
+                self.slot = current_slot;
+
+                self.counter.is_rate_limited(client, limit, quantity, quota)
+            }
+            Ordering::Equal => self.counter.is_rate_limited(client, limit, quantity, quota),
+        }
+    }
+}
+
+impl Default for SlottedCounter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Represents the local budget taken from a global quota.
+struct Counter {
+    local_counter: LocalCounter,
+    redis_counter: RedisCounter,
+}
+
+impl Counter {
+    pub fn new() -> Self {
+        Self {
+            local_counter: LocalCounter::new(),
+            redis_counter: RedisCounter::new(),
+        }
+    }
+
+    pub fn is_rate_limited(
+        &mut self,
+        client: &mut PooledClient,
+        limit: u64,
+        quantity: usize,
+        quota: &RedisQuota,
+    ) -> Result<bool, RedisError> {
+        if self.local_counter.try_consume(quantity) {
+            return Ok(false);
+        }
+
+        if !self.redis_counter.can_satisfy(quantity, limit) {
+            return Ok(true);
+        }
+
+        let budget_to_reserve = quantity.max(self.default_request_size_based_on_limit());
+        let reserved = self
+            .redis_counter
+            .try_reserve(client, budget_to_reserve, limit, quota)? as usize;
+
+        self.local_counter.increase_budget(reserved);
+        Ok(!self.local_counter.try_consume(quantity))
+    }
+
+    fn default_request_size_based_on_limit(&self) -> usize {
+        100
```

Comment:
We need to actually implement this, use a percentage we grab from the config

Reply:
Implemented with a global constant for now.

Might in the future want to use some kind of moving average instead.